File: android/src/main/kotlin/com/cloudwebrtc/webrtc/controller/VideoRendererFactoryController.kt
package com.cloudwebrtc.webrtc.controller

import com.cloudwebrtc.webrtc.FlutterRtcVideoRenderer
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.TextureRegistry

/**
 * Controller for creating new [FlutterRtcVideoRenderer]s.
 *
 * @property binaryMessenger messenger used for creating new [MethodChannel]s.
 * @property textureRegistry registry with which new textures will be created.
 */
class VideoRendererFactoryController(
    private val binaryMessenger: BinaryMessenger,
    private val textureRegistry: TextureRegistry
) :
    MethodChannel.MethodCallHandler {
    /**
     * Channel which will be listened for the [MethodCall]s.
     */
    private val methodChannel =
        MethodChannel(binaryMessenger, ChannelNameGenerator.name("VideoRendererFactory", 0))

    init {
        methodChannel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "create" -> {
                val renderer = FlutterRtcVideoRenderer(textureRegistry)
                result.success(VideoRendererController(binaryMessenger, renderer).asFlutterResult())
            }
        }
    }
}

File: ios/Classes/controller/VideoRendererFactoryController.swift
import Flutter

/// Controller which will create new `FlutterRtcVideoRenderer`s.
class VideoRendererFactoryController {
  /// Flutter messenger for creating channels.
  private var messenger: FlutterBinaryMessenger

  /// Method channel for communicating with Flutter side.
  private var channel: FlutterMethodChannel

  /// Flutter texture registry for creating new Flutter textures.
  private var registry: FlutterTextureRegistry

  /// Creates new `VideoRendererFacotyrController`,
  init(messenger: FlutterBinaryMessenger, registry: FlutterTextureRegistry) {
    let channelName = ChannelNameGenerator.name(name: "VideoRendererFactory", id: 0)
    self.messenger = messenger
    self.registry = registry
    self.channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
    self.channel.setMethodCallHandler({ (call, result) in
      try! self.onMethodCall(call: call, result: result)
    })
  }

  /**
    Handles all support Flutter method calls.

    Creates new `FlutterRtcVideoRenderer`s.
  */
  func onMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) throws {
    switch call.method {
    case "create":
      let renderer = FlutterRtcVideoRenderer(registry: self.registry)
      result(
        VideoRendererController(messenger: self.messenger, renderer: renderer).asFlutterResult())
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
